xarray/core/coordinates.py

```diff
@@ -796,6 +797,126 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(DatasetCoordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+    """
+
+    # TODO: This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    # and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    # This will add both the coordinates variables and their index.
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
+
+    _data: DataTree
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        variables = self._data._data_variables | self._data._coord_variables
+        return Frozen({k: v for k, v in variables.items() if k in self._names})
+
+    def __getitem__(self, key: Hashable) -> DataArray:
+        if key in self._data._data_variables or key in self._data.children:
+            raise KeyError(key)
+        return self._data[key]
+
+    def to_dataset(self) -> Dataset:
+        """Convert these coordinates into a new Dataset"""
+        return self._data.to_dataset()._copy_listed(self._names)
+
+    def _update_coords(
+        self, coords: dict[Hashable, Variable], indexes: Mapping[Any, Index]
+    ) -> None:
```

Comment:
I would consider implementing this with a two part approach:

1. Create the new node dataset
2. Check alignment (using the utility from datatree.py)
3. Assign the dataset properties to the existing tree

Something like (untested):
```python
# create updated node
node_ds = self._data.to_dataset(inherited=False)
node_ds.coords._update_coords(coords, indexes)
# check consistency
if self._data.parent is not None:
    parent_ds = self._data.parent._to_dataset_view(rebuild_dims=False)
    datatree._check_alignment(self._data.path, node_ds, self._data.children)
# assign updated attributes
coord_variables = {k: v for k, v in node_ds.variables.items() if k in node_ds._coord_names}
self._data._node_coord_variables = coord_variables
self._data._node_dims = node_ds.dims
self._data._indexes = node_ds._indexes
```

Reply:
This is helpful - I've refactored using this approach in https://github.com/pydata/xarray/pull/9451/commits/7ecdd16018b8b7b75f94ad531112b9572d165ba4

Reply:
But it's exposed some kind of bug with `dt.copy`, see https://github.com/pydata/xarray/issues/9454.

Reply:
Okay now that #9454 was resolved by #9457, I've successfully reimplemented this class using your suggested approach.